crates/compiler/src/obj_gen/mod.rs

```diff
@@ -2545,73 +2781,138 @@ impl ObjectGenerator {
+            return Ok(());
+        }
+
+        // To determine how to call the function, we need to pull out the argument and
+        // return types.
+        let arg_types = function_arguments
             .iter()
             .map(|op| LLVMType::try_from(op.get_type()))
             .collect::<StdResult<Vec<_>>>()?;
-        let return_type = LLVMType::try_from(instruction.get_type())?;
+        let ret_type = LLVMType::try_from(instruction.get_type())?;
+        let func_type = LLVMFunction::new(ret_type.clone(), arg_types.as_slice());
 
-        // If it is NOT one of these, we have to actually generate the call, and to do
-        // that we need to know whether it is an internal or external call.
-        let function_ref = if let Some(id) = func_ctx.module_functions().get(&function_name) {
-            BlockRef::Local(*id)
-        } else if let Some(polyfill_name) =
-            self.polyfills
-                .get_polyfill(&function_name, &operand_types, &return_type)
-        {
-            BlockRef::Builtin(polyfill_name.to_string())
+        // We also need our arguments and return values, remembering to register the
+        // latter in the context.
+        let mut function_inputs = function_arguments
+            .iter()
+            .map(|op| util::get_var_or_const(op, bb, func_ctx))
+            .collect::<Result<Vec<_>>>()?;
+        let output_type_flo = ObjectContext::flo_type_of(&ret_type)?;
+        let output_var = bb.add_variable(output_type_flo.clone());
+
+        // The name for the output variable may not exist depending on where this call
+        // comes from, so we may have to create one.
+        let output_name = if let Some(name) = instruction.get_name() {
+            name.to_str()?.to_string()
         } else {
-            BlockRef::External(function_name)
+            self.name_supply.borrow_mut().allocate()
         };
+        func_ctx.register_local(output_var, &output_name, output_type_flo);
 
-        // We then need to put together the operands by grabbing the variables from the
-        // context and passing them in. Given that the function name is the LAST
-        // operand, it is the one we want to ignore.
-        let inputs = value_operands
-            .into_iter()
-            .map(|op| util::get_var_or_const(&op, bb, func_ctx))
-            .collect::<Result<Vec<_>>>()?;
+        let function_outputs =
+            vec![bb.add_variable(ObjectContext::flo_type_of(&ret_type.clone())?)];
 
-        // We also the need to handle providing the return value, if one exists, using
-        // the type and whether the instruction has a name telling us.
-        let outputs = util::make_opcode_output(&instruction, bb, func_ctx)?;
+        // To call the function we need to resolve the block reference through which it
+        // will be called.
+        let func_ref = if let Some(id) = func_ctx.module_functions().get(&func_name) {
```

Comment:
Is there a possibility of a name-collision, here? I'm trying to convince myself there isn't, but my head is a bit foggy at the moment.

These names come from the LLVM IR, if I'm parsing correctly, and are forced to be either declared (e.g. externs) or defined (in this module's scope) -- so the LLVM emitter is effectively responsible for making sure there's no name clash between those otherwise-unnamespace'd names?

Reply:
Yeah, these names come from IR, so the LLVM _ingester_ will perform a check to avoid name clashes. Added a comment along these lines.